File: Lunchtime/Model/Location/LunchtimeGeocoder.m
//
//  LunchtimeGeocoder.m
//  Lunchtime
//
//  Created by Willow Belle on 2015-11-15.
//  Copyright © 2015 Cosmic Labs. All rights reserved.
//

#import "LunchtimeGeocoder.h"

@implementation LunchtimeGeocoder

- (void)reverseGeocodeLocationWithCoordinate:(CLLocationCoordinate2D)coordinate withCompletionHandler:(void (^)(CLPlacemark *placemark))completionHandler {

    CLLocation *location = [[CLLocation alloc] initWithLatitude:coordinate.latitude longitude:coordinate.longitude];
    CLGeocoder *geocoder = [CLGeocoder new];

    [geocoder reverseGeocodeLocation:location completionHandler:^(NSArray<CLPlacemark *> *placemarks, NSError *error) {
        if (error) {
            NSLog(@"Reverse Geocoding Location Error: %@", error);

            return;
        }

        NSArray *placeArray = placemarks;
        CLPlacemark *placemark = placeArray[0];

        completionHandler(placemark);
    }];
}

- (void)geocodeLocationWithAddress:(NSString *)address withCompletionHandler:(void (^)(CLLocationDegrees latitude, CLLocationDegrees longitude))completionHandler {

    CLGeocoder *geocoder = [CLGeocoder new];

    [geocoder geocodeAddressString:address completionHandler:^(NSArray<CLPlacemark *> *placemarks, NSError *error) {
        if (error) {
            NSLog(@"Geocoding Location Error: %@", error);

            return;
        }

        CLPlacemark *placemark = placemarks[0];
        completionHandler(placemark.location.coordinate.latitude, placemark.location.coordinate.longitude);
    }];
}

@end

File: Lunchtime/Model/Location/LunchtimeGeocoder.swift
//
//  LunchtimeGeocoder.swift
//  Lunchtime
//
//  Created by Willow Bumby on 2015-11-19.
//  Copyright © 2015 Lighthouse Labs. All rights reserved.
//

import UIKit

public class LunchtimeGeocoder: NSObject {

    class func reverseGeocodeRequestWithCoordinate(coordinate: CLLocationCoordinate2D, handler: ((placemark: CLPlacemark) -> Void)) {
        let location = CLLocation(latitude: coordinate.latitude, longitude: coordinate.longitude)
        let geocoder = CLGeocoder()

        geocoder.reverseGeocodeLocation(location) { (placemarks, error) in
            guard let placemarks = placemarks else {
                return
            }

            let placemark = placemarks[0]
            handler(placemark: placemark)
        }
    }

    class func geocodeRequestWithAddress(address: String, handler: ((latitude: CLLocationDegrees, longitude: CLLocationDegrees) -> Void)) {
        let geocoder = CLGeocoder()

        geocoder.geocodeAddressString(address, completionHandler: { (placemarks, error) in
            guard error != nil else {
                print("Geocode request failed with \(error)")

                return
            }

            guard let placemarks = placemarks else {
                print("The placemarks array is nil")

                return
            }

            guard let coordinate = placemarks[0].location?.coordinate else {
                print("The coordinate struct is nil")

                return
            }

            handler(latitude: coordinate.latitude, longitude: coordinate.longitude)
        })
    }
}